File: 986.interval-list-intersections.java
import java.util.List;

/*
 * @lc app=leetcode id=986 lang=java
 *
 * [986] Interval List Intersections
 */
class Solution {
    public int[][] intervalIntersection(int[][] A, int[][] B) {
        List<int[]> res = new ArrayList();
        int i = 0, j = 0;
        while (i < A.length && j < B.length) {
            int low = Math.max(A[i][0], B[j][0]);
            int high = Math.min(A[i][1], B[j][1]);
            if (low <= high)
                res.add(new int[] { low, high });
            if (A[i][1] < B[j][1])
                i++;
            else
                j++;

        }

        int[][] result = res.toArray(new int[0][]);

        return result;
    }
}

File: 986.interval-list-intersections.js
/*
 * @lc app=leetcode id=986 lang=javascript
 *
 * [986] Interval List Intersections
 */
/**
 * @param {number[][]} A
 * @param {number[][]} B
 * @return {number[][]}
 */

// Approach 1: Merge Intervals O(M+N) O(M+N)
var intervalIntersection = function (A, B) {
  let ans = [];
  let i = 0,
    j = 0;
  while (i < A.length && j < B.length) {
    let lo = Math.max(A[i][0], B[j][0]);
    let hi = Math.min(A[i][1], B[j][1]);
    if (lo <= hi) {
      ans.push([lo, hi]);
    }

    if (A[i][1] < B[j][1]) i++;
    else j++;
  }

  return ans;
};

// intervalIntersection(
//   [
//     [0, 2],
//     [5, 10],
//     [13, 23],
//     [24, 25],
//   ],
//   [
//     [1, 5],
//     [8, 12],
//     [15, 24],
//     [25, 26],
//   ]
// );
